File: src/main/java/com/github/frcsty/frozenactions/actions/player/CenterMessageAction.java
package com.github.frcsty.frozenactions.actions.player;

import com.github.frcsty.frozenactions.actions.Action;
import com.github.frcsty.frozenactions.actions.ActionContext;
import com.github.frcsty.frozenactions.util.DefaultFontInfo;
import com.github.frcsty.frozenactions.util.Extensions;
import org.jetbrains.annotations.NotNull;

public final class CenterMessageAction implements Action {

    private static final int CENTER_PX = 154;
    private static final int MAX_PX = 250;

    @Override
    public @NotNull String getId() {
        return "CENTERMESSAGE";
    }

    @Override
    public void run(final @NotNull ActionContext context) {
        execute(context);
    }

    public static void execute(final ActionContext context) {
        String message = Extensions.colorize(context.getData());
        int messagePxSize = 0;

        boolean previousCode = false;
        boolean isBold = false;

        int charIndex = 0;
        int lastSpaceIndex = 0;
        String toSendAfter = null;
        String recentColorCode = "";

        for (char character : message.toCharArray()) {
            if (character == '\u00a7') {
                previousCode = true;
                continue;
            } else if (previousCode) {
                previousCode = false;
                recentColorCode = "§" + character;
                if (character == 'l' || character == 'L') {
                    isBold = true;
                    continue;
                } else {
                    isBold = false;
                }
            } else if (character == ' ') {
                lastSpaceIndex = charIndex;
            } else {
                final DefaultFontInfo dfi = DefaultFontInfo.getDefaultFontInfo(character);

                messagePxSize += isBold ? dfi.getBoldLength() : dfi.getLength();
                messagePxSize++;
            }

            if (messagePxSize >= MAX_PX) {
                toSendAfter = recentColorCode + message.substring(lastSpaceIndex + 1);
                message = message.substring(0, lastSpaceIndex + 1);
                break;
            }

            charIndex++;
        }

        final int halvedMessageSize = messagePxSize / 2;
        final int toCompensate = CENTER_PX - halvedMessageSize;
        final int spaceLength = DefaultFontInfo.SPACE.getLength() + 1;

        int compensated = 0;
        final StringBuilder builder = new StringBuilder();
        while (compensated < toCompensate) {
            builder.append(" ");
            compensated += spaceLength;
        }

        context.getPlayer().sendMessage(builder.toString() + message);
        if (toSendAfter != null) {
            execute(new ActionContext(context.getPlugin(), context.getPlayer(), toSendAfter));
        }
    }
}

File: src/main/kotlin/com/github/frcsty/frozenactions/actions/player/CenterMessageAction.kt
package com.github.frcsty.frozenactions.actions.player

import com.github.frcsty.frozenactions.actions.Action
import com.github.frcsty.frozenactions.util.DefaultFontInfo
import com.github.frcsty.frozenactions.util.color
import org.bukkit.entity.Player

object CenterMessageAction : Action {

    private const val CENTER_PX = 154
    private const val MAX_PX = 250
    override val id = "CENTERMESSAGE"

    override fun run(player: Player, data: String) {
        var message = data.color()
        var messagePxSize = 0

        var previousCode = false
        var isBold = false

        var charIndex = 0
        var lastSpaceIndex = 0
        var toSendAfter: String? = null
        var recentColorCode = ""

        for (c in message.toCharArray()) {
            if (c == '§') {
                previousCode = true
                continue
            } else if (previousCode) {
                previousCode = false
                recentColorCode = "§$c"
                if (c == 'l' || c == 'L') {
                    isBold = true
                    continue
                } else {
                    isBold = false
                }
            } else if (c == ' ') {
                lastSpaceIndex = charIndex
            } else {
                val dfi: DefaultFontInfo = DefaultFontInfo.getDefaultFontInfo(c)
                messagePxSize += if (isBold) dfi.boldLength else dfi.length
                messagePxSize++
            }
            if (messagePxSize >= MAX_PX) {
                toSendAfter = recentColorCode + message.substring(lastSpaceIndex + 1)
                message = message.substring(0, lastSpaceIndex + 1)
                break
            }
            charIndex++
        }

        val halvedMessageSize = messagePxSize / 2
        val toCompensate: Int = CENTER_PX - halvedMessageSize
        val spaceLength: Int = DefaultFontInfo.SPACE.length + 1
        var compensated = 0

        val sb = StringBuilder()
        while (compensated < toCompensate) {
            sb.append(" ")
            compensated += spaceLength
        }

        player.sendMessage(sb.toString() + message)
        if (toSendAfter != null) {
            run(player, toSendAfter)
        }
    }
}